nucypher/datastore/datastore.py

```diff
@@ -14,269 +14,222 @@
+        `record_type` identified by the `record_id` provided.
+
+        When `writeable` is `False`, the record returned by this method
+        cannot be used for any operations that write to the datastore. If an
+        attempt is made to retrieve a non-existent record whilst `writeable`
+        is `False`, this method raises a `RecordNotFound` error.
+
+        When `writeable` is `True`, the record can be used to perform writes
+        on the datastore. In the event an error occurs during the write, the
+        transaction will be aborted and no data will be written, and a
+        `DatastoreTransactionError` will be raised.
+
+        If the record is used outside the scope of the context manager, any
+        writes or reads will error.
+        """
+        with suppress(ValueError):
+            # If the ID can be converted to an int, we do it.
+            record_id = int(record_id)
+
+        with self.__db_env.begin(write=writeable) as datastore_tx:
+            record = record_type(datastore_tx, record_id, writeable=writeable)
+            try:
+                yield record
+            except (AttributeError, TypeError, DBWriteError) as tx_err:
+                # Handle `RecordNotFound` cases when `writeable` is `False`.
+                if not writeable and isinstance(tx_err, AttributeError):
+                    raise RecordNotFound(tx_err)
+                raise DatastoreTransactionError(f'An error was encountered during the transaction (no data was written): {tx_err}')
+            finally:
+                # Now we ensure that the record is not writeable
+                record.__dict__['_DatastoreRecord__writeable'] = False
+
+    @contextmanager
+    def query_by(self,
+              record_type: 'DatastoreRecord',
+              filter_func: Callable[[Union[Any, 'DatastoreRecord']], bool]=None,
+              filter_field: str="",
+              writeable: bool=False,
+              ) -> List['DatastoreRecord']:
+        """
+        Performs a query on the datastore for the record by `record_type`.
+
+        An optional `filter_func` callable will take the decoded field
+        specified by the optional arg `filter_field` (see below) for the given
+        `record_type` iff the `filter_field` has been provided.
+        If no `filter_field` has been provided, then the `filter_func` will
+        receive a _readonly_ `DatastoreRecord`.
+
+        An optional `filter_field` can be provided as a `str` to perform a
+        query on a specific field for a `record_type`. This will cause the
+        `filter_func` to receive the decoded `filter_field` per the `record_type`.
+        Additionally, providing a `filter_field` will limit the query to
+        iterating over only the subset of records specific to that field.
+
+        If records can't be found, this method will raise `RecordNotFound`.
+        """
+        valid_records = set()
+        with self.__db_env.begin(write=writeable) as datastore_tx:
```

Comment:
This `with` block is massive!

Reply:
Yes, it's a bit unfortunate. I'd say it's mostly comments. :P
I'd try to make it more condense, but this is the API that `lmdb` offers.

Reply:
The size of this with block is not a limitation of lmbd. You can break up this function into smaller ones.

Reply:
I tried that previously. Imho, it makes the method as a whole much less readable.